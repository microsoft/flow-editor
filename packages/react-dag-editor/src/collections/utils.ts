export const is = Object.is;

export class MapIterator<T, U> implements IterableIterator<U> {
  private readonly upstream: Iterator<T>;
  private readonly f: (value: T) => U;

  public constructor(upstream: Iterator<T>, f: (value: T) => U) {
    this.upstream = upstream;
    this.f = f;
  }

  public [Symbol.iterator](): IterableIterator<U> {
    return this;
  }

  public next(): IteratorResult<U> {
    const next = this.upstream.next();
    if (next.done) {
      return next;
    }
    return {
      done: false,
      value: this.f(next.value)
    };
  }
}
