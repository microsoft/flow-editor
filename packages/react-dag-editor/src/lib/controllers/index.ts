export * from "./DragController";
export * from "./DragNodeController";
export * from "./TouchController";
export * from "./TwoFingerHandler";
