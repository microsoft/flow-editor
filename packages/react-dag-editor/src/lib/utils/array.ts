import { is } from "./is";

export function mapCow<T, U>(
  list: readonly T[],
  f: (value: T, index: number) => U
): U[] {
  const values = [] as U[];
  let same = true;
  for (let i = 0; i < list.length; i += 1) {
    const value0 = list[i];
    const value = f(value0, i);
    same = same && is(value0, value);
    values.push(value);
  }
  return same ? ((list as unknown) as U[]) : values;
}
