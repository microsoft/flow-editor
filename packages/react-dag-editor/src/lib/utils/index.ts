export * from "./getPortPosition";
export * from "./getNodeConfig";
export * from "./graphDataUtils";
export * from "./transformMatrix";
export * from "./getCurvePathD";
export * from "./layout";
export * from "./browsers";
export * from "./isWithinThreshold";
export * from "./geometric";
export * from "./viewport";
export * from "./Writable";
export * from "./gantt/getCriticalPath";
export * from "./gantt/getTopoSortingNodes";
export * from "./getOffsetLimit";
export * from "./event";
export * from "./valueOrCallback";
export * from "./debug";
export * from "./debounce";
export * from "./arrayEquals";
export * from "./empty";
export * from "./stopPropagation";
export * from "./clamp";
export * from "./history";
export * as Bitset from "./bitset";
