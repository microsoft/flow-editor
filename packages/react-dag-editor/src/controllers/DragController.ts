/* eslint-disable no-invalid-this */
import { IEventProvider, IGlobalMoveEventTypes } from "../event-provider/types";
import { IPoint } from "../models/geometry";
import { noop } from "../utils/noop";
import { animationFramed, IScheduledCallback } from "../utils/scheduling";

export type TGetPositionFromEvent = (e: MouseEvent) => IPoint;

export interface IOnDragMove<T = MouseEvent> {
  /**
   * clientX after getPositionFromEvent
   */
  clientX: number;
  /**
   * clientY after getPositionFromEvent
   */
  clientY: number;
  dx: number;
  dy: number;
  totalDX: number;
  totalDY: number;
  e: T;
}

export interface IOnDragEnd<T = MouseEvent> {
  totalDX: number;
  totalDY: number;
  e: T;
}

export class DragController<T extends MouseEvent = MouseEvent> {
  public onMove: (move: IOnDragMove) => void = noop;
  public onEnd: (end: IOnDragEnd) => void = noop;
  protected mouseMove: IScheduledCallback<[T]>;
  protected lastEvent: T | null = null;
  private readonly getPositionFromEvent: TGetPositionFromEvent;
  private startX = 0;
  private startY = 0;
  private prevClientX = 0;
  private prevClientY = 0;
  private readonly eventProvider: IEventProvider<IGlobalMoveEventTypes>;

  public constructor(
    eventProvider: IEventProvider<IGlobalMoveEventTypes>,
    getPositionFromEvent: TGetPositionFromEvent
  ) {
    this.eventProvider = eventProvider;
    this.getPositionFromEvent = getPositionFromEvent;
    this.mouseMove = animationFramed((e: T) => {
      this.doOnMouseMove(e);
    });
  }

  public start(e: T): void {
    this.lastEvent = e;
    const { x, y } = this.getPositionFromEvent(e);
    this.startX = x;
    this.startY = y;
    this.prevClientX = x;
    this.prevClientY = y;

    this.eventProvider.on("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.on("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  public stop(): void {
    this.mouseMove.cancel();
    this.eventProvider.off("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.off("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  protected getDelta(x: number, y: number): IPoint {
    const clientDX = x - this.prevClientX;
    const clientDY = y - this.prevClientY;
    this.prevClientX = x;
    this.prevClientY = y;
    return {
      x: clientDX,
      y: clientDY,
    };
  }

  protected getTotalDelta(e: MouseEvent): IPoint {
    const x = e.clientX - this.startX;
    const y = e.clientY - this.startY;
    return {
      x,
      y,
    };
  }

  protected doOnMouseMove(e: MouseEvent): void {
    const { x, y } = this.getPositionFromEvent(e);
    const { x: dx, y: dy } = this.getDelta(x, y);
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onMove({
      clientX: x,
      clientY: y,
      dx,
      dy,
      totalDX,
      totalDY,
      e,
    });
  }

  protected doOnMouseUp(e: T): void {
    e.preventDefault();
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onEnd({ totalDX, totalDY, e });

    this.stop();
  }

  private readonly onMouseUp = (e: T) => {
    this.lastEvent = e;
    this.doOnMouseUp(e);
    this.lastEvent = null;
  };

  private readonly onMouseMove = (e: T) => {
    this.lastEvent = e;
    e.preventDefault();
    this.mouseMove(e);
  };
}

export function defaultGetPositionFromEvent<T extends MouseEvent>(
  e: T
): IPoint {
  return {
    x: e.clientX,
    y: e.clientY,
  };
}
