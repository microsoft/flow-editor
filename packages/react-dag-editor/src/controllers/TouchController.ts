/* eslint-disable no-invalid-this */
import * as React from "react";
import { animationFramed } from "../utils/scheduling";

export interface ITouchHandler<ExtraArgs extends unknown[] = []> {
  onStart(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void;
  onMove(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void;
  onEnd(e: PointerEvent, ...args: ExtraArgs): void;
}

export class TouchController<ExtraArgs extends unknown[] = []> {
  public readonly eventHandlers = {
    onPointerDown: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers = new Map(this.pointers);
      this.pointers.set(e.pointerId, e.nativeEvent);
      this.updateHandler(e.nativeEvent, ...args);
    },

    onPointerMove: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers.set(e.pointerId, e.nativeEvent);
      this.onMove(e.nativeEvent, ...args);
    },

    onPointerUp: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers = new Map(this.pointers);
      this.pointers.delete(e.pointerId);
      this.updateHandler(e.nativeEvent, ...args);
    }
  };

  private pointers = new Map<number, PointerEvent>();
  private currentHandler: ITouchHandler<ExtraArgs> | undefined;
  private readonly handlers: Map<number, ITouchHandler<ExtraArgs>>;

  private readonly onMove = animationFramed(
    (e: PointerEvent, ...args: ExtraArgs) => {
      this.currentHandler?.onMove(this.pointers, e, ...args);
    }
  );

  public constructor(handlers: Map<number, ITouchHandler<ExtraArgs>>) {
    this.handlers = handlers;
  }

  private updateHandler(e: PointerEvent, ...args: ExtraArgs): void {
    const handler = this.handlers.get(this.pointers.size);
    if (handler !== this.currentHandler) {
      this.currentHandler?.onEnd(e, ...args);
      this.currentHandler = handler;
      this.currentHandler?.onStart(this.pointers, e, ...args);
    }
  }
}
