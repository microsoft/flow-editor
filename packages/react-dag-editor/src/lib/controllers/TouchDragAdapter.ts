import { IEventProvider, IGlobalMoveEventTypes } from "../event-provider/types";
import { $Cons } from "../utils/tuple";
import { ITouchHandler } from "./TouchController";

export class TouchDragAdapter<ExtraArgs extends unknown[] = []>
  implements
    IEventProvider<IGlobalMoveEventTypes<ExtraArgs>>,
    ITouchHandler<ExtraArgs> {
  private readonly startListeners = new Set<
    (...args: $Cons<MouseEvent, ExtraArgs>) => void
  >();
  private readonly moveListeners = new Set<
    (...args: $Cons<MouseEvent, ExtraArgs>) => void
  >();
  private readonly endListeners = new Set<
    (...args: $Cons<MouseEvent, ExtraArgs>) => void
  >();

  public off(
    type: "start" | "move" | "end",
    callback: (...args: $Cons<MouseEvent, ExtraArgs>) => void
  ): TouchDragAdapter<ExtraArgs> {
    switch (type) {
      case "start":
        this.startListeners.delete(callback);
        break;
      case "move":
        this.moveListeners.delete(callback);
        break;
      case "end":
        this.endListeners.delete(callback);
        break;
      default:
    }
    return this;
  }

  public on(
    type: "start" | "move" | "end",
    callback: (...args: $Cons<MouseEvent, ExtraArgs>) => void
  ): TouchDragAdapter<ExtraArgs> {
    switch (type) {
      case "start":
        this.startListeners.add(callback);
        break;
      case "move":
        this.moveListeners.add(callback);
        break;
      case "end":
        this.endListeners.add(callback);
        break;
      default:
    }
    return this;
  }

  public onStart(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void {
    this.startListeners.forEach(cb => {
      cb.call(undefined, e, ...args);
    });
  }

  public onMove(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void {
    this.moveListeners.forEach(cb => {
      cb.call(undefined, e, ...args);
    });
  }

  public onEnd(e: PointerEvent, ...args: ExtraArgs): void {
    this.endListeners.forEach(cb => {
      cb.call(undefined, e, ...args);
    });
  }
}
