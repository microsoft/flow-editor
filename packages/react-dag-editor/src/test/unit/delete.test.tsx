import { act, render } from "@testing-library/react";
import * as React from "react";
import {
  dataReadonlyMode,
  GraphCanvasEvent,
  GraphEdgeStatus,
  GraphModel,
  GraphNodeStatus,
  GraphPortStatus,
  notSelected,
} from "../../index";
import { GraphController } from "../../lib/controllers/GraphController";
import { GraphControllerRef, TestComponent } from "../TestComponent";
import { makeEdges, makeNodes, makePorts } from "../utils";

const ports = makePorts([
  GraphPortStatus.Default,
  GraphPortStatus.Default,
  GraphPortStatus.Default,
]);

const deleteSelectedItems = (data: GraphModel) => {
  return data.deleteItems({
    node: notSelected,
    edge: notSelected,
  });
};

it("should do nothing if delete disabled", () => {
  const data = GraphModel.fromJSON({
    edges: makeEdges([
      [GraphEdgeStatus.ConnectedToSelected, ["0", "0"], ["1", "1"]],
      [GraphEdgeStatus.UnconnectedToSelected, ["0", "1"], ["2", "0"]],
    ]),
    nodes: makeNodes(
      [
        GraphNodeStatus.ConnectedToSelected,
        GraphNodeStatus.Selected,
        GraphNodeStatus.UnconnectedToSelected,
      ],
      ports
    ),
  });

  const graphControllerRef = React.createRef<GraphController>();
  render(
    <TestComponent
      data={data}
      settings={{
        features: dataReadonlyMode,
      }}
    >
      <GraphControllerRef ref={graphControllerRef} />
    </TestComponent>
  );

  const graphController = graphControllerRef.current!;
  expect(graphController).toBeDefined();

  act(() => {
    graphController.dispatch({
      type: GraphCanvasEvent.Delete,
    });
  });

  expect(graphController.state.data.present.toJSON()).toMatchSnapshot();
});

it("should do nothing if nothing selected", () => {
  const data = GraphModel.fromJSON({
    edges: makeEdges([
      [GraphEdgeStatus.Default, ["0", "0"], ["1", "1"]],
      [GraphEdgeStatus.Default, ["0", "1"], ["2", "0"]],
    ]),
    nodes: makeNodes(
      [
        GraphNodeStatus.Default,
        GraphNodeStatus.Editing,
        GraphNodeStatus.Default,
      ],
      ports
    ),
  });

  expect(deleteSelectedItems(data).toJSON()).toMatchSnapshot();
});

it("should delete related edges", () => {
  const data = GraphModel.fromJSON({
    edges: makeEdges([
      [GraphEdgeStatus.ConnectedToSelected, ["0", "0"], ["1", "1"]],
      [GraphEdgeStatus.UnconnectedToSelected, ["0", "1"], ["2", "0"]],
    ]),
    nodes: makeNodes(
      [
        GraphNodeStatus.ConnectedToSelected,
        GraphNodeStatus.Selected,
        GraphNodeStatus.UnconnectedToSelected,
      ],
      ports
    ),
  });

  expect(deleteSelectedItems(data).toJSON()).toMatchSnapshot();
});

it("should only delete node", () => {
  const data = GraphModel.fromJSON({
    edges: [],
    nodes: makeNodes(
      [
        GraphNodeStatus.UnconnectedToSelected,
        GraphNodeStatus.Selected,
        GraphNodeStatus.UnconnectedToSelected,
      ],
      ports
    ),
  });

  expect(deleteSelectedItems(data).toJSON()).toMatchSnapshot();
});

it("should only delete edge", () => {
  const data = GraphModel.fromJSON({
    edges: makeEdges([
      [GraphEdgeStatus.Selected, ["0", "0"], ["1", "1"]],
      [GraphEdgeStatus.Default, ["0", "1"], ["2", "0"]],
    ]),
    nodes: makeNodes(
      [
        GraphNodeStatus.Default,
        GraphNodeStatus.Default,
        GraphNodeStatus.Default,
      ],
      ports
    ),
  });
  expect(deleteSelectedItems(data).toJSON()).toMatchSnapshot();
});
