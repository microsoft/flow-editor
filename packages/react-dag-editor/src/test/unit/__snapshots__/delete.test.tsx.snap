// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should delete related edges 1`] = `
Object {
  "edges": Array [
    Object {
      "automationId": undefined,
      "id": "1",
      "properties": Object {},
      "shape": undefined,
      "source": "0",
      "sourcePortId": "1",
      "status": 0,
      "target": "2",
      "targetPortId": "0",
    },
  ],
  "nodes": Array [
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "0",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "2",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
  ],
}
`;

exports[`should delete related edges: events 1`] = `Array []`;

exports[`should do nothing if delete disabled 1`] = `
Object {
  "edges": Array [
    Object {
      "automationId": undefined,
      "id": "0",
      "properties": Object {},
      "shape": undefined,
      "source": "0",
      "sourcePortId": "0",
      "status": 4,
      "target": "1",
      "targetPortId": "1",
    },
    Object {
      "automationId": undefined,
      "id": "1",
      "properties": Object {},
      "shape": undefined,
      "source": "0",
      "sourcePortId": "1",
      "status": 8,
      "target": "2",
      "targetPortId": "0",
    },
  ],
  "nodes": Array [
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "0",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 8,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "1",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 1,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "2",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 16,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
  ],
}
`;

exports[`should do nothing if delete disabled: events 1`] = `
Array [
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
]
`;

exports[`should do nothing if nothing selected 1`] = `
Object {
  "edges": Array [
    Object {
      "automationId": undefined,
      "id": "0",
      "properties": Object {},
      "shape": undefined,
      "source": "0",
      "sourcePortId": "0",
      "status": 0,
      "target": "1",
      "targetPortId": "1",
    },
    Object {
      "automationId": undefined,
      "id": "1",
      "properties": Object {},
      "shape": undefined,
      "source": "0",
      "sourcePortId": "1",
      "status": 0,
      "target": "2",
      "targetPortId": "0",
    },
  ],
  "nodes": Array [
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "0",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "1",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 4,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "2",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
  ],
}
`;

exports[`should do nothing if nothing selected: events 1`] = `Array []`;

exports[`should only delete edge 1`] = `
Object {
  "edges": Array [
    Object {
      "automationId": undefined,
      "id": "1",
      "properties": Object {},
      "shape": undefined,
      "source": "0",
      "sourcePortId": "1",
      "status": 0,
      "target": "2",
      "targetPortId": "0",
    },
  ],
  "nodes": Array [
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "0",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "1",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "2",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
  ],
}
`;

exports[`should only delete edge: events 1`] = `Array []`;

exports[`should only delete node 1`] = `
Object {
  "edges": Array [],
  "nodes": Array [
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "0",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
    Object {
      "ariaLabel": undefined,
      "automationId": undefined,
      "height": undefined,
      "id": "2",
      "name": undefined,
      "ports": Array [
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "0",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "0",
          "position": Array [
            0.25,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "1",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "1",
          "position": Array [
            0.5,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "id": "2",
          "isInputDisabled": undefined,
          "isOutputDisabled": undefined,
          "name": "2",
          "position": Array [
            0.75,
            0,
          ],
          "properties": Object {},
          "shape": undefined,
          "status": 0,
        },
      ],
      "properties": Object {},
      "shape": undefined,
      "status": 0,
      "width": undefined,
      "x": 0,
      "y": 0,
    },
  ],
}
`;

exports[`should only delete node: events 1`] = `Array []`;
